### PR TITLE
CHAIN-340 telegram mini app milestones

### PR DESCRIPTION
**Pull request by imykolenko**
=== iPhone 12 mini - scroll appears when all components visible ===

 

=== Desktop ===
Screenshot 2024-07-01 at 1 06 45 PM Screenshot 2024-07-01 at 1 01 35 PM

### CODE REVIEW

File: backend/src/main/kotlin/co/chainring/core/model/db/migrations/V53_TelegramMiniAppMilestones.kt [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import co.chainring.core.model.telegram.miniapp.TelegramMiniAppUserId
+import org.jetbrains.exposed.sql.SchemaUtils
+import org.jetbrains.exposed.sql.kotlin.datetime.timestamp
+import org.jetbrains.exposed.sql.transactions.transaction
+
+@Suppress("ClassName")
+class V53_TelegramMiniAppMilestones : Migration() {
+
+    object V53_TelegramMiniAppUserTable : GUIDTable<TelegramMiniAppUserId>("telegram_mini_app_user", ::TelegramMiniAppUserId) {
+        val invites = long("invites").default(0)
```
Comment (bholzman) [inline backend/src/main/kotlin/co/chainring/core/model/db/migrations/V53_TelegramMiniAppMilestones.kt:14]: I think I didn't mention this, but the mythos spec calls for each user to start with 5 invites, so we could probably change this default to that

File: telegram-mini-app/src/components/MainScreen/HomeTab.tsx [hunk 2]
```diff
@@ -76,17 +79,34 @@ export default function HomeTab({ user }: { user: User }) {
...
+          user.lastMilestone.grantedAt.getTime() + fiveMinutesInMs >
+            new Date().getTime() && (
+            <div className="my-2 flex flex-col items-center justify-center">
+              <MilestoneReachedWidget milestone={user.lastMilestone} />
+            </div>
+          )}
+        {user.nextMilestoneIn && (
+          <div className="my-2 flex flex-col items-center justify-center">
+            <div className="text-center text-white">
+              Next milestone in {user.nextMilestoneIn.toFixed(0)} CR!
```
Comment (bholzman) [inline telegram-mini-app/src/components/MainScreen/HomeTab.tsx:105]: maybe format with commas for > 3 digits?

**Review by bholzman — APPROVED** [review]

File: backend/src/main/kotlin/co/chainring/core/model/db/migrations/V53_TelegramMiniAppMilestones.kt (hunk 1 above)
Comment (imykolenko) [reply backend/src/main/kotlin/co/chainring/core/model/db/migrations/V53_TelegramMiniAppMilestones.kt]: sure, made 5 default

File: telegram-mini-app/src/apiClient.ts [hunk 3]
```diff
@@ -57,11 +57,20 @@ const CheckInStreakSchema = z.object({
 })
 export type CheckInStreak = z.infer<typeof CheckInStreakSchema>
 
+const LastMilestoneSchema = z.object({
+  invites: z.number(),
+  grantedAt: z.coerce.date()
+})
+export type LstMilestone = z.infer<typeof LastMilestoneSchema>
```
Comment (antonyshchenko) [inline telegram-mini-app/src/apiClient.ts:64]: Nit: typo

**Pull request merged**